product_multi_ean/models/product_product.py

```diff
@@ -93,16 +93,16 @@ def _prepare_ean13_vals(self):
 
     @api.model
     def search(self, domain, *args, **kwargs):
-        if list(filter(lambda x: x[0] == 'barcode', domain)):
-            ean_operator = list(
-                filter(lambda x: x[0] == 'barcode', domain)
-            )[0][1]
-            ean_value = list(
-                filter(lambda x: x[0] == 'barcode', domain)
-            )[0][2]
-
-            eans = self.env['product.ean13'].search(
-                [('name', ean_operator, ean_value)])
-            domain = list(filter(lambda x: x[0] != 'barcode', domain))
-            domain += [('ean13_ids', 'in', eans.ids)]
+        for sub_domain in list(filter(lambda x: x[0] == 'barcode', domain)):
+            domain = self._get_ean13_domain(sub_domain, domain)
         return super(ProductProduct, self).search(domain, *args, **kwargs)
+
+    def _get_ean13_domain(self, sub_domain, domain):
+        ean_operator = sub_domain[1]
+        ean_value = sub_domain[2]
+        eans = self.env['product.ean13'].search(
+            [('name', ean_operator, ean_value)])
+        domain = [('ean13_ids', 'in', eans.ids)
+                  if x[0] == 'barcode' and x[2] == ean_value
```

Comment:
This is redundant

Reply:
Maybe but it keeps proper domain format across more complex filters.

Before this patch if you look for multiple products you only get ean_ids for one of them, with this change I'm ensuring that even if redundancy or not, search domains are full of it, search has consistency.

Reply:
I mean you are already checking this before calling this method.

Reply:
@pedrobaeza the loop is on domain, but the filter is on subdomain. Right @hveficent ?

Reply:
Yes, here i go across the domain looking for specific subdomain that needs to be changed. Without this if there is a multiple search, it will overwrite both 'barcode' domains. I also need to check that domain contains barcode because in most common search case by product it look for name, internal reference and barcode

Reply:
Uf, the code is a bit confusing. Why not using `osv.expression.OR` method for doing this in one shot?

Reply:
It could be an option, I never used before, so I think we can move this forward to fix current issue, and maybe I will create another PR to improve readability using OR expression.

Reply:
OK, I'm not blocking this PR, but I don't want also to approve without understanding full code, and now I'm afraid that I don't have time for deeper analysis, so go ahead if you think it's correct.

Reply:
:+1: